scripts/license/add_license_header.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	rootDir, err := os.Getwd()
+	if err != nil {
+		log.Fatalf("Failed to get working directory: %v\n", err)
+	}
+	rootDir = filepath.Join(rootDir, "../..")
+
+	// Process files with specified extensions
+	result := 0
+	for ext, prefix := range extensions {
+		fmt.Printf("Processing files with extension %s using prefix '%s'\n", ext, prefix)
+		err := processFiles(rootDir, ext, prefix, string(license), *checkOnly, *checkDoubleHeader)
+		if err != nil {
+			log.Fatalf("Error processing files with extension %s: %v\n", ext, err)
+		}
+	}
+	os.Exit(result)
+}
+
+func processFiles(root, ext, prefix, license string, checkOnly, doubleHeader bool) error {
+	licenseHeader := addPrefix(license, prefix)
+	var files []string
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil || info.IsDir() {
+			return nil
+		}
+		if shouldIgnore(path) {
+			return nil
+		}
+		if matchExtension(path, ext) {
+			files = append(files, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to walk directory %s: %v", root, err)
+	}
+	for _, f := range files {
+		if doubleHeader {
+			if err := checkDoubleHeader(f, prefix); err != nil {
+				fmt.Println(err)
+			}
+			continue
+		}
+		if err := processFile(f, licenseHeader, checkOnly); err != nil {
+			fmt.Println(err)
+			if checkOnly {
+				continue
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func processFile(path, licenseHeader string, checkOnly bool) error {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return fmt.Errorf("failed to read %s: %v", path, err)
+	}
+	// we check the first 20 lines because the license header should be the top 15 lines.
+	lines := strings.SplitN(string(content), "\n", 20)
+	licenseLines := strings.Split(strings.TrimSuffix(licenseHeader, "\n"), "\n")
+	needsUpdate := false
+
+	// check if the file has the first line of geth lincense header
+	if strings.Contains(lines[0], "The go-ethereum Authors") {
+		// license header is 15 lines, 16th should be empty
+		if strings.TrimSpace(lines[15]) != "" {
+			lines[14] += "\n"
+		}
+		return nil
+	}
+
+	for i, l := range licenseLines {
+		if i >= len(lines) || strings.TrimSpace(lines[i]) != strings.TrimSpace(l) {
+			needsUpdate = true
+			break
+		}
+	}
+	if !needsUpdate {
+		return nil
+	}
+	if checkOnly {
+		return fmt.Errorf("missing or incorrect license header: %s", path)
+	}
+
+	// this means the file has an old license header, we need to replace it
+	if strings.Contains(lines[0], "Sonic Operations Ltd") {
+		// search for the first empty line after the old license header
+		counter := 0
+		for i, line := range lines {
+			counter += len(line)
+			if strings.TrimSpace(line) == "" {
+				// remove lines up to this point
+				content = []byte(strings.Join(lines[i+1:], "\n"))
+				break
+			}
+		}
+	}
+
+	// Add header
+	newContent := licenseHeader + "\n" + string(content)
+	return os.WriteFile(path, []byte(newContent), 0222)
```

Comment:
Is this script implicitly changing the permission of all my files to public writeable? And no read permissions?

It would be good to preserve the same file permissions. Wdyt?

Reply:
https://github.com/0xsoniclabs/sonic/pull/335/commits/3c89a0aec3e33779aed3ff7afe2bea5b3fbb7a75